transports/quic/src/transport.rs

```diff
@@ -191,18 +214,65 @@ impl<P: Provider> Transport for GenTransport<P> {
                 &mut listeners[index].dialer_state
             }
         };
-        Ok(dialer_state.new_dial(socket_addr, self.handshake_timeout, version))
+        Ok(dialer_state.new_dial(socket_addr, handshake_timeout, version))
     }
 
     fn dial_as_listener(
         &mut self,
         addr: Multiaddr,
     ) -> Result<Self::Dial, TransportError<Self::Error>> {
-        // TODO: As the listener of a QUIC hole punch, we need to send a random UDP packet to the
-        // `addr`. See DCUtR specification below.
-        //
-        // https://github.com/libp2p/specs/blob/master/relay/DCUtR.md#the-protocol
-        Err(TransportError::MultiaddrNotSupported(addr))
+        let (socket_addr, _version, peer_id) = self.remote_multiaddr_to_socketaddr(addr.clone())?;
+
+        let peer_id = peer_id.ok_or(TransportError::MultiaddrNotSupported(addr))?;
+
+        let listeners = self.eligible_listeners(&socket_addr);
+
+        let endpoint_channel = match listeners.as_slice() {
+            [] => {
+                return Err(TransportError::Other(
+                    Error::NoActiveListenerForDialAsListener,
+                ));
+            }
+            [listener] => listener.endpoint_channel.clone(),
+            listeners => {
+                // Pick any listener to use for dialing.
+                // We hash the socket address to achieve determinism.
+                let mut hasher = DefaultHasher::new();
+                socket_addr.hash(&mut hasher);
+                let index = hasher.finish() as usize % listeners.len();
+                listeners[index].endpoint_channel.clone()
+            }
+        };
```

Comment:
Can we merge this into `self.eligible_listeners` if we return `Option<&mut Listener>`? This code can then extract the `endpoint_channel` from it and use `ok_or` to handle the `None` case.

The other usage can `match` in the return value, create a new dialer in the `None` case and extract the `dialer_state` in the `Some` case.